### PR TITLE
Fixes #1645 by introducing ENFORCE_ITEMS_IMMUTABILITY

### LINKED ISSUE #1645: Serialization behavior differs between test runs
This might be related to #1536 opened 4 weeks ago.

**Edit: See post below for updated details**

A unit test creates two grains **g1** and **g2**, where **g2** is connected to a stream populated by **g1**. This stream transfers objects of type `MyObject`. `MyObject` contains one event that is marked as `[field: NonSerialized]`. 

When **g1** is sending objects of type `MyObject` to the stream sometimes the events are still connected and sometimes not, while they should always be disconnected.

Is this an issue with TestingSiloHost or a general issue (**Update**: It is a general issue)? This inconsistent behavior makes Orleans unusable for my purposes right now. Let me know if I can be of any help fixing that issue.

Reproduction project for 1.1.3 is available at https://github.com/bwanner/Orleans-Event-Serialization

### PR DESCRIPTION
**Pull request by bwanner**
Adding an attribute `EnforceItemsImmutability` for SimpleMessageStream with a default value of `true`. Also adding a unit test to validate this behavior.

### CODE REVIEW

File: test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs [hunk 1]
```diff
@@ -372,6 +374,54 @@ private void Heading(string testName)
...
+            await itemProducer.ProduceTestObject(SMS_STREAM_PROVIDER_NAME_NO_IMMUTABILITY);
+
+            Assert.IsFalse(await consumerSameSilo.TestObjectWasTransferredImmutable());
+            Assert.IsFalse(await consumerDifferentSilo.TestObjectWasTransferredImmutable());
+
+            await consumerSameSilo.UnsubscribeFromStream();
+            await consumerDifferentSilo.UnsubscribeFromStream();
+
+            // Test behavior if immutability is enabled
+            itemProducer = GrainClient.GrainFactory.GetGrain<IStreamingImmutabilityTestGrain>(Guid.NewGuid());
```
Comment (bwanner) [inline test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs]: Usage of `Immutable` stays disabled if I reuse the existing itemProducer. Any idea why this happens or is this intended behavior?

**Comment by gabikliot** [comment]
Lets please change the default value to false, to preserve backward compatibility.
We can have a separate discussion about changing the default permanently.

File: src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs [hunk 2]
```diff
@@ -27,6 +29,10 @@ public Task Init(string name, IProviderRuntime providerUtilitiesManager, IProvid
             string fireAndForgetDeliveryStr;
             fireAndForgetDelivery = config.Properties.TryGetValue(FIRE_AND_FORGET_DELIVERY, out fireAndForgetDeliveryStr) && Boolean.Parse(fireAndForgetDeliveryStr);
 
+            string enforceItemsImmutabilityStr;
+            config.Properties.TryGetValue(ENFORCE_ITEMS_IMMUTABILITY, out enforceItemsImmutabilityStr);
+            enforceItemsImmutability = (enforceItemsImmutabilityStr == null) ? true : Boolean.Parse(enforceItemsImmutabilityStr);
```
Comment (gabikliot) [inline src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs]: the default value should be defined in one place. I already saw 3.

File: src/Orleans/Streams/Internal/IStreamGrainExtensions.cs [hunk 3]
```diff
@@ -11,6 +11,7 @@ namespace Orleans.Streams
     internal interface IStreamConsumerExtension : IGrainExtension
     {
         Task<StreamHandshakeToken> DeliverItem(GuidId subscriptionId, Immutable<object> item, StreamSequenceToken currentToken, StreamHandshakeToken handshakeToken);
```
Comment (gabikliot) [inline src/Orleans/Streams/Internal/IStreamGrainExtensions.cs]: Try giving different names to those 2 method: DeliverMutable and DeliveeImmutable.

**Comment by bwanner** [comment]
@gabikliot If the default value is set to true existing behavior is preserved. Current behavior was to always wrap items in an Immutable.

File: src/Orleans/Streams/Internal/StreamConsumerExtension.cs [hunk 4]
```diff
@@ -67,13 +67,18 @@ internal bool RemoveObserver(GuidId subscriptionId)
...
-        public Task<StreamHandshakeToken> DeliverItem(GuidId subscriptionId, Immutable<object> item, StreamSequenceToken currentToken, StreamHandshakeToken handshakeToken)
+        public Task<StreamHandshakeToken> DeliverImmutable(GuidId subscriptionId, Immutable<object> item, StreamSequenceToken currentToken, StreamHandshakeToken handshakeToken)
         {
-            if (logger.IsVerbose3) logger.Verbose3("DeliverItem {0} for subscription {1}", item.Value, subscriptionId);
+            return DeliverMutable(subscriptionId, item.Value, currentToken, handshakeToken);
+        }
+
+        public Task<StreamHandshakeToken> DeliverMutable(GuidId subscriptionId, object item, StreamSequenceToken currentToken, StreamHandshakeToken handshakeToken)
+        {
+            if (logger.IsVerbose3) logger.Verbose3("DeliverItem {0} for subscription {1}", item, subscriptionId);
```
Comment (jason-bragg) [inline src/Orleans/Streams/Internal/StreamConsumerExtension.cs]: I understand that you're just preserving the existing logging, but not knowing the type of item, i'd be very hesitant to log the actual item.

File: src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs [hunk 5]
```diff
@@ -12,20 +12,31 @@ public class SimpleMessageStreamProvider : IInternalStreamProvider
...
             providerRuntime = (IStreamProviderRuntime) providerUtilitiesManager;
+
             string fireAndForgetDeliveryStr;
-            fireAndForgetDelivery = config.Properties.TryGetValue(FIRE_AND_FORGET_DELIVERY, out fireAndForgetDeliveryStr) && Boolean.Parse(fireAndForgetDeliveryStr);
+            config.Properties.TryGetValue(FIRE_AND_FORGET_DELIVERY, out fireAndForgetDeliveryStr);
+            fireAndForgetDelivery = (fireAndForgetDeliveryStr == null) ? DEFAULT_VALUE_FIRE_AND_FORGET_DELIVERY : Boolean.Parse(fireAndForgetDeliveryStr);
+
+            string enforceItemsImmutabilityStr;
+            config.Properties.TryGetValue(ENFORCE_ITEMS_IMMUTABILITY, out enforceItemsImmutabilityStr);
+            enforceItemsImmutability = (enforceItemsImmutabilityStr == null) ? DEFAULT_VALUE_ENFORCE_ITEMS_IMMUTABILITY : Boolean.Parse(enforceItemsImmutabilityStr);
```
Comment (jason-bragg) [inline src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProvider.cs]: Suggest using ProviderConfigurationExtensions.GetBoolProperty

File: test/TestInternalGrains/StreamingImmutabilityTestGrain.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            if (_streamSubscriptionHandle != null)
+                await _streamSubscriptionHandle.UnsubscribeAsync();
+        }
+
+        public async Task ProduceTestObject(string providerName)
+        {
+            _myObject = new StreamImmutabilityTestObject() {MyString = this.GetPrimaryKey().ToString()};
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(this.GetPrimaryKey(), "Namespace");
+            await stream.OnNextAsync(_myObject);
```
Comment (jason-bragg) [inline test/TestInternalGrains/StreamingImmutabilityTestGrain.cs]: Immutable means that the application level promises not to modify the data, so the stream system can optimize around this promise and avoid data copies (faster).  To test if this is happening, you'd need the application level to break this promise by modifying the data and see if the sent data is similarly modified.

All that is tested here is that the object was send/received, not whether it was immutable or not.  To test that the immutable setting was applied, you must violate the implicit promise not to mutate the state.  Something like:
Task sendTask = streamOnNextAsync(_myObject)
_myObject.MyString = "blarg"
return sendTask.
Then in onNextAsync check to see if the received object's string is this.GetPrimaryKey().ToString() or "Blarg".   If it is "Blarg", then the stream was set to immutable, because the original state of the object when the send was made was not copied.

File: test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs (hunk 1 above)
Comment (jason-bragg) [inline test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs]: May be a bug.  Producer code looks like it get's a stream from the specified provider.  We do cache the producer objects but that should be done by fully qualified stream id, which includes the provider name, so you shouldn't need to get a new itemProducers here.

File: test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs [hunk 7]
```diff
@@ -372,6 +374,54 @@ private void Heading(string testName)
...
         }
 
+        public async Task StreamTest_22_TestImmutabilityDuringStreaming()
+        {
+            Heading("StreamTest_22_TestImmutabilityDuringStreaming");
+
+            IStreamingImmutabilityTestGrain itemProducer = GrainClient.GrainFactory.GetGrain<IStreamingImmutabilityTestGrain>(Guid.NewGuid());
+            string producerSilo = await itemProducer.GetSiloIdentifier();
+
+            // Setup consumers one in silo of item producer and the other one in another silo.
```
Comment (jason-bragg) [inline test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs]: I don't understand why silo matters?  If a producer and consumer are local and stream is immutable, the same object is transferred, whereas remotely a new object is created, but the difference between immutable calls is whether there is a data copy made at call time, in which silo location is not relevant.

File: test/TestInternalGrains/StreamingImmutabilityTestGrain.cs (hunk 6 above)
Comment (bwanner) [inline test/TestInternalGrains/StreamingImmutabilityTestGrain.cs]: You're right, immutability includes two aspects in Orleans, of which I just tested aspect 1:
1. Immutable objects are transferred as is within the same silo and serialized between different silos.
2. As you mentioned application level promises not to modify the data.

I went with testing only aspect 1 because I wanted to make sure that disabling of immutability leads to correct serialization of `StreamImmutabilityTestObject`, so `MyString` is not transferred.

If there is need I can go and test aspect 2, but it should result from aspect 1 intentionally.

File: test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs (hunk 7 above)
Comment (bwanner) [inline test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs]: Silo locality plays a role in the current implementation where all objects got wrapped as immutable.

If an item is transferred from a producer P to a local consumer CL and a remote consumer CR, then P and CL are accessing the same item while CR is only able to access a copy. So you have immutable items, that are not globally immutable.

The added behavior by setting `ENFORCE_ITEMS_IMMUTABILITY = false` allows to always deal with copies, even if producer and consumer are within the same silo. The fact that the item has been copied is validated via the emptyness of the `StreamImmutabilityTestObject.MyString` which is marked as `[NonSerialized]`.

File: test/TestInternalGrains/StreamingImmutabilityTestGrain.cs (hunk 6 above)
Comment (jason-bragg) [inline test/TestInternalGrains/StreamingImmutabilityTestGrain.cs]: > "two aspects" 

Yes and no.  The immutability alters exactly one behavior, weather a deep copy of an object is made at call time.  All other 'aspects' are side effects of our RPC pattern.

This means what you're testing is side effects, not the actual behavior.  As these side effects depend on a great many other factors (like grain locality) these tests could fail (in the future) due to perfectly reasonable changes to other parts of the system.

I think I now understand how you are approaching this, but I don't think testing undocumented and unsupported side effects rather than the specifics of the feature is a good approach.

This is a good improvement, so I feel bad about being such a stickler on the testing, but these test harnesses are public and are often examined by those trying to understand the system.  This test not only test side effects, but would also give readers a false impression of what immutability means.

File: test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs (hunk 7 above)
Comment (jason-bragg) [inline test/TesterInternal/StreamingTests/SingleStreamTestRunner.cs]: I missed the [NonSerialized] part of the test the first pass.  Thanks for the  clarification. I think I now understand how the test works.

File: src/Orleans/Streams/Internal/StreamConsumerExtension.cs (hunk 4 above)
Comment (jason-bragg) [inline src/Orleans/Streams/Internal/StreamConsumerExtension.cs]: Maybe truncate item string to something reasonable? 128 characters?

File: test/TestInternalGrains/StreamingImmutabilityTestGrain.cs (hunk 6 above)
Comment (bwanner) [inline test/TestInternalGrains/StreamingImmutabilityTestGrain.cs]: > Yes and no. The immutability alters exactly one behavior, weather a deep copy of an object is made at call time. All other 'aspects' are side effects of our RPC pattern.

I think I just fully understood how that immutability is used in the context of multiplexing. Immutability does not make sense for a remote consumer nor should it change anything.
Would you want to change the test similar to what you mentioned in your first comment while differentiating between a local and remote consumer?
Or would you prefer to just test the immutability behavior with a local consumer and not consider the remote consumer case (now my preferred option)?

It's not stickling, hurray for code and test quality :+1:

File: src/Orleans/Streams/Internal/StreamConsumerExtension.cs (hunk 4 above)
Comment (bwanner) [inline src/Orleans/Streams/Internal/StreamConsumerExtension.cs]: Sounds like a compromise. Knowing about all contents of the objects might be useful for debugging, but just increases log size too much (even if its Verbose3).

File: test/TestInternalGrains/StreamingImmutabilityTestGrain.cs (hunk 6 above)
Comment (jason-bragg) [inline test/TestInternalGrains/StreamingImmutabilityTestGrain.cs]: > Or would you prefer to just test the immutability behavior with a local consumer and not consider the remote consumer case

I think this is the simplest test which gets at the heart of the behavior.  If we make an OnNextAsync call, then immediately change the object state before resolving the task, the delivered event (regardless of locality) should have the original value when immutability is off, and the modified value when immutability is on.

**Comment by gabikliot** [comment]
I think we need to change the config value name.
ENFORCE IMMUTABILITY means for me: runtime is enforcing immutability, by deep copying. The default is NOT to deep copy, by wrapping in Immutable attribute. Wrapping in Immutable attribute is NOT enforcing, it is the contrary: it is Assuming (relaying on app to guarantee immutability), NOT Enforcing.

Better to change the name to something like: DEEP_COPY_ITEMS

**Comment by bwanner** [comment]
I don't fully get how immutability can be enforced by deep copying? From my understanding the opposite is the case. An object being enforced as immutable means for me that it's wrapped in an Immutable and thus not deep copied.

I don't know if I would use the term "deep copy" in the setting, according to what @jason-bragg said it seems to be an implementation aspect of immutability:
> The immutability alters exactly one behavior, weather a deep copy of an object is made at call time.

How about: WRAP_ITEMS_IMMUTABLE = true for naming and default behavior?

**Comment by gabikliot** [comment]
I will let @jason-bragg decide on that.

**Comment by jason-bragg** [comment]
TL;DR - I suggest "OptimizeForImmutableData"

>Immutable means that the application level promises not to modify the data, so the stream system can optimize around this promise and avoid data copies.

Usually when an application passes an immutable argument to a grain call, it is attesting that the data is immutable.  Orleans takes advantage of this fact to optimize delivery by not doing deep copies, but does not enforce the immutability itself.  [Using Immutable to Optimize Copying](http://dotnet.github.io/orleans/Advanced-Concepts/Using-Immutable-to-Optimize-Copying)

My preference would be to have SMS not do any special handling of the data, and require service developers to call the streaming APIs with Immutable<T> if they want immutable data, but we already have this capability, so I think we are kinda stuck with it.  That being the case, we need to communicate this behavior in a way that indicates that we are not making the data immutable, but simply optimizing for immutable data, as we do for grain calls.  Hence the setting name OptimizeForImmutableData.

**Comment by jason-bragg** [comment]
Looks good.  Plz squash and rebase and we'll merge.

**Comment by bwanner** [comment]
Done :)
